ibis/formats/parser.py

```diff
@@ -0,0 +1,213 @@
+    DataType.Type.SMALLINT: dt.int16,
+    DataType.Type.SMALLMONEY: dt.int32,
+    DataType.Type.TEXT: dt.string,
+    DataType.Type.TIME: dt.time,
+    DataType.Type.TIMETZ: dt.time,
+    DataType.Type.TINYINT: dt.int8,
+    DataType.Type.UBIGINT: dt.uint64,
+    DataType.Type.UINT: dt.uint32,
+    DataType.Type.USMALLINT: dt.uint16,
+    DataType.Type.UTINYINT: dt.uint8,
+    DataType.Type.UUID: dt.uuid,
+    DataType.Type.VARBINARY: dt.binary,
+    DataType.Type.VARCHAR: dt.string,
+    DataType.Type.VARIANT: dt.json,
+    DataType.Type.UNIQUEIDENTIFIER: dt.uuid,
+    #############################
+    # Unsupported sqlglot types #
+    #############################
+    # BIGSERIAL = auto()
+    # DATETIME64 = auto() # clickhouse
+    # ENUM = auto()
+    # INT4RANGE = auto()
+    # INT4MULTIRANGE = auto()
+    # INT8RANGE = auto()
+    # INT8MULTIRANGE = auto()
+    # NUMRANGE = auto()
+    # NUMMULTIRANGE = auto()
+    # TSRANGE = auto()
+    # TSMULTIRANGE = auto()
+    # TSTZRANGE = auto()
+    # TSTZMULTIRANGE = auto()
+    # DATERANGE = auto()
+    # DATEMULTIRANGE = auto()
+    # HLLSKETCH = auto()
+    # IMAGE = auto()
+    # IPPREFIX = auto()
+    # ROWVERSION = auto()
+    # SERIAL = auto()
+    # SET = auto()
+    # SMALLSERIAL = auto()
+    # SUPER = auto()
+    # TIMESTAMPLTZ = auto()
+    # UNKNOWN = auto()  # Sentinel value, useful for type annotation
+    # UINT128 = auto()
+    # UINT256 = auto()
+    # USERDEFINED = "USER-DEFINED"
+    # XML = auto()
+}
+
+
+class TypeParser(Singleton, Slotted):
+    @property
+    @abc.abstractmethod
+    def dialect(self) -> str:
+        """The dialect this parser is for."""
+
+    default_nullable = True
+    """Default nullability when not specified."""
+
+    default_decimal_precision: int | None = None
+    """Default decimal precision when not specified."""
+
+    default_decimal_scale: int | None = None
+    """Default decimal scale when not specified."""
+
+    default_temporal_scale: int | None = None
+    """Default temporal scale when not specified."""
+
+    default_interval_precision: str | None = None
+    """Default interval precision when not specified."""
+
+    fallback: Mapping[str, dt.DataType] = FrozenDict()
+    """Default fallback mapping of SQL string types to ibis types."""
+
+    default_fallback: Mapping[str, dt.DataType] = FrozenDict(
+        {
+            "INTERVAL YEAR TO MONTH": dt.Interval("M"),
+            "INTERVAL DAY TO SECOND": dt.Interval("ms"),
+        }
+    )
+    """Mapping of SQL string types to ibis types shared by all parser child classes."""
+
+    @property
+    def _fallback(self) -> Mapping[str, dt.DataType]:
+        return FrozenDict({**self.default_fallback, **self.fallback})
+
+    def parse(self, text: str) -> dt.DataType:
+        """Parse a type string into an ibis data type."""
+        fallback = self._fallback
+        if dtype := fallback.get(text, fallback.get(text.upper())):
```

Comment:
It seems to be a bit counterintuitive to me, if there is a fallback then return with it before parsing?

Reply:
Yes. It's a way to workaround missing support for a given type in sqlglot.

Reply:
I mean I'd expect to return with the fallback value after failing to parse the input.

Reply:
This is used to support the various specific geometry types in postgres like `line` and `point`.

Reply:
So IIUC if there is an exact typestring match in either of the fallback mappings this will return with that exact ibis type, otherwise it will use sqlglot to do the actual parsing. 

Could we use a different naming here instead of `fallback`? That indicates a different behaviour to me.

Reply:
Changed to `short_circuit`